plugins/algoltek-usbcr/fu-algoltek-usbcr-device.c

```diff
@@ -0,0 +1,671 @@
+							regs[i].val,
+							FU_AG_USBCR_RD_WR_XDATA,
+							error))
+			return FALSE;
+	}
+	do {
+		if (!fu_algoltek_usbcr_device_read_reg(self,
+						       0xC8,
+						       buf,
+						       sizeof(buf),
+						       FU_AG_USBCR_RD_WR_XDATA,
+						       error))
+			return FALSE;
+	} while (buf[0] & 0x01);
+
+	/* success*/
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usbcr_device_command_wrsr(FuAlgoltekUsbcrDevice *self, gboolean en, GError **error)
+{
+	guint8 buf[1] = {0};
+	FuAgUsbcrRegSetup regs[] = {
+	    {0xC8, 0x04},
+	    {0xCA, 0x01},
+	    {0x400, FU_AG_USBCR_WRSR},
+	    {0x401, 0x00},
+	    {0xC9, 0x02},
+	    {0xC8, 0x05},
+	};
+	if (en == FU_AG_BLOCK_MODE_EN)
+		regs[3].val = 0x0C;
+	for (guint i = 0; i < G_N_ELEMENTS(regs); i++) {
+		if (!fu_algoltek_usbcr_device_write_reg(self,
+							regs[i].reg,
+							regs[i].val,
+							FU_AG_USBCR_RD_WR_XDATA,
+							error))
+			return FALSE;
+	}
+	do {
+		if (!fu_algoltek_usbcr_device_read_reg(self,
+						       0xC8,
+						       buf,
+						       sizeof(buf),
+						       FU_AG_USBCR_RD_WR_XDATA,
+						       error))
+			return FALSE;
+	} while (buf[0] & 0x01);
+
+	/* success*/
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usbcr_device_command_rdsr(FuAlgoltekUsbcrDevice *self, GError **error)
+{
+	guint8 buf[1] = {0};
+	FuAgUsbcrRegSetup regs[] = {
+	    {0xC8, 0x04},
+	    {0xCA, 0x01},
+	    {0x400, FU_AG_USBCR_RDSR},
+	    {0xC9, 0x01},
+	    {0xC4, 0x01},
+	    {0xC7, 0x00},
+	    {0xC8, 0x07},
+	};
+	for (guint i = 0; i < G_N_ELEMENTS(regs); i++) {
+		if (!fu_algoltek_usbcr_device_write_reg(self,
+							regs[i].reg,
+							regs[i].val,
+							FU_AG_USBCR_RD_WR_XDATA,
+							error))
+			return FALSE;
+	}
+	do {
+		if (!fu_algoltek_usbcr_device_read_reg(self,
+						       0xC8,
+						       buf,
+						       sizeof(buf),
+						       FU_AG_USBCR_RD_WR_XDATA,
+						       error))
+			return FALSE;
+	} while (buf[0] & 0x01);
+
+	/* success*/
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usbcr_device_spi_flash_block_mode_cb(FuDevice *device,
+						 gpointer user_data,
+						 GError **error)
+{
+	FuAlgoltekUsbcrDevice *self = FU_ALGOLTEK_USBCR_DEVICE(device);
+	guint8 buf[2] = {0};
+	guint8 en = GPOINTER_TO_INT(user_data);
+
+	/* set command wren */
+	if (!fu_algoltek_usbcr_device_command_wren(self, error))
+		return FALSE;
+
+	/* set command wrsr */
+	if (!fu_algoltek_usbcr_device_command_wrsr(self, en, error))
+		return FALSE;
+
+	/* set command rdsr */
+	if (!fu_algoltek_usbcr_device_command_rdsr(self, error))
+		return FALSE;
+
+	/* read data */
+	if (!fu_algoltek_usbcr_device_read_reg(self,
+					       0x400,
+					       buf,
+					       sizeof(buf),
+					       FU_AG_USBCR_RD_WR_XDATA,
+					       error))
+		return FALSE;
+	if (en == FU_AG_BLOCK_MODE_DISEN) {
+		if ((buf[0] & 0xC) != 0x0) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_DATA,
+				    "The return value is 0x%x while expecting value is 0x0",
```

Comment:
GError messages shouldn't be Title Case -- maybe delete the `The ` completely.

Reply:
Updated.